graphql_compiler/compiler/emit_sql.py

```diff
@@ -74,31 +77,249 @@ def _find_columns_used(sql_schema_info, ir):
+        """Set output columns for the fold object."""
+        self._outputs = outputs
+
+    def _set_group_by(self, group_by):
+        """Set output columns for the fold object."""
+        self._group_by = group_by
+
+    def _construct_fold_joins(self, edge, from_alias, to_alias):
+        """Use the edge descriptors to create the join clause between the tables in the fold."""
+        join_clause = sqlalchemy.join(
+            from_alias,
+            to_alias,
+            onclause=(from_alias.c[edge.from_column] == to_alias.c[edge.to_column])
+        )
+        return join_clause
+
+    def _construct_fold_subquery(self, subquery_from_clause):
+        """Combine all parts of the fold object to produce the complete fold subquery."""
+        return sqlalchemy.select(
+            self.outputs
+        ).select_from(
+            subquery_from_clause
+        ).group_by(
+            *self.group_by
+        )
+
+    def _get_fold_outputs(self, fold_scope_location, join_descriptor, all_folded_outputs):
+        """Generate output columns for innermost fold scope and add them to active SQLFoldObject."""
+        # find outputs for this fold in all_folded_outputs and add to self._outputs
+        if fold_scope_location.fold_path in all_folded_outputs:
+            for fold_output in all_folded_outputs[fold_scope_location.fold_path]:
+                # distinguish between folds with the same fold path but different query paths
+                if (fold_output.base_location, fold_output.fold_path) == (
+                        fold_scope_location.base_location, fold_scope_location.fold_path):
+                    if fold_output.field == '_x_count':
+                        raise NotImplementedError(u'_x_count not implemented in SQL')
+
+                    # force column to have explicit label as opposed to anon_label
+                    intermediate_fold_output_name = 'fold_output_' + fold_output.field
+                    # add array aggregated output column to self._outputs
+                    self._outputs.append(
+                        sqlalchemy.func.array_agg(
+                            self.output_vertex_alias.c[fold_output.field]
+                        ).label(intermediate_fold_output_name)
+                    )
+
+        # this is the unique identifier for the outer vertex used to join to the outer table
+        self._outputs.append(self.outer_vertex_alias.c[join_descriptor.from_column])
+
+        return sorted(self._outputs, key=lambda column: column.name, reverse=True)
+
+    def visit_output_vertex(self,
+                            output_alias,
+                            fold_scope_location,
+                            join_descriptor,
+                            all_folded_outputs):
+        """Update output columns when visiting the vertex containing output directives."""
+        self._output_vertex_alias = output_alias
+        self._outputs = self._get_fold_outputs(fold_scope_location,
```

Comment:
we should probably raise an error here if a user tries to call `visit_output_vertex` more than once.